compiler/sem/sem.nim

```diff
@@ -735,6 +735,8 @@ proc semAfterMacroCall(c: PContext, call, macroResult: PNode,
         let err = newError(c.config, result, PAstDiag(kind: adSemExpressionHasNoType))
         localReport(c.config, err)
         result = newSymNode(errorSym(c, result, err))
+      elif typ.kind == tyError and typ.n != nil and typ.n.kind == nkError:
+        result = typ.n
```

Comment:
Not particularly happy about this, as there are plenty of other usages of `semTypeNode` that could also have the same sort of issue. It's not clear if `semTypeNode` should do a `localReport` or callsites should check things. They all seem flawed in their own way.

Reply:
I think you could use `semTypeNode2`, which is meant for the type-AST/normal-AST edge, for the whole `tyTypeDesc` branch, like so:
```nim
      if result.kind == nkStmtList: result.transitionSonsKind(nkStmtListType)
      result = semTypeNode2(c, result)
      if result.kind != nkError:
        result.typ = makeTypeDesc(c, result.typ)
```
The error for the "no type" case would be a different one, `adSemTypeExpected` vs. `adSemExpressionHasNoType`, although I'd argue that the former is more correct, since here a *type expression* is expected rather than just a *typed expression*.

`semTypeNode2` seems to be missing the `typ.n != nil` check, however -- that'd need to be fixed.

Reply:
Thanks, that's a fair bit cleaner. I can fix any tests that break due to message changes, and I agree that the message is more correct.